### ISSUE TITLE
How to handle faliures in subscribing?

### ISSUE THREAD

**Issue by mgritzmacher** [bytebeamio/rumqtt#206]
Hi. Thanks for creating this library. I'm running into some problems trying to add error handling with my use of the rumqttc client.

When I tell the client to subscribe to something I know the topic, but I don't know what pkid the client will use
```rust
client.subscribe(topic, QoS::AtLeastOnce).await
```
when I get an Event for an outgoing packet, I only have the pkid
```
Ok(Outgoing(Subscribe(4)))
```
when I get an acknowledgement, I get the pkid and if each topic was successful or unsucessful, but I've lost the topic
```
Ok(Incoming(SubAck(SubAck { pkid: 4, return_codes: [Success(AtLeastOnce)] })))
```

I can see that the Subscribe type for Incoming packets https://docs.rs/rumqttc/0.3.0/rumqttc/struct.Subscribe.html has the topic information and pkid in presumably the same order as for Incoming SubAck, but there's no matching topic information on the outgoing Subscribe variant https://docs.rs/rumqttc/0.3.0/rumqttc/enum.Outgoing.html#variant.Subscribe

I want to connect the information together so that I can try to resubscribe to topics that failed, but I can't see how to do this with the public API of the library at the moment.

**Comment by tekjar** [comment]
Yeah I think it makes sense to add subscription filters to `Outgoing::Subscribe`. But this can lead to the rabbit hole of adding topics for `Outgoing::Publish` and some other metadata for other packets.

I think the deeper problem is that the client doesn't have synchronous mode yet. Does your subscriptions happen during initialization or is it a complicated logic? If it's during initialization, you can simulate synchronous client like this

```
client.subscribe();
for event in connection.iter() {
    /// wait until suback packet
}
```

Few other folks asked for synchronous mode before. I'll probably implement it in January

**Comment by mgritzmacher** [comment]
In my particular use case I'm using the client asyncronously and my subscriptions are quite complicated. I've got a many to one relationship with services that tell the client to do things, and a central service that polls the event loop, so trying to control the order of things (in order to know which ack corresponds to which subscribe) would be challenging.

**Comment by tekjar** [comment]
Ok. I'll add filters to `Outgoing::Subscribe` for now. Let me know if you want to implement this. This should be straight forward

**Comment by mgritzmacher** [comment]
I've made a pull request that works for my needs with error handling, let me know if it needs any adjustments.
